File: visualizer/frontend/src/lib/toolPolicy.ts
import type { DiscoveryPhase } from './discoveryPhaseMachine';

export interface ToolPolicy {
  allowed: string[];
  blocked: string[];
}

export const toolPolicyForPhase = (phase: DiscoveryPhase): ToolPolicy => {
  switch (phase) {
    case 'residue':
      return {
        allowed: ['inspect_residues', 'get_residue_state', 'set_metric', 'highlight_residues'],
        blocked: ['extract_pockets', 'map_pharmacophore_features', 'screen_fragments', 'run_docking_surrogate', 'generate_report'],
      };
    case 'topology':
      return {
        allowed: ['search_pocket_vectors', 'inspect_topology', 'graph_metrics', 'highlight_clusters', 'map_to_structure'],
        blocked: ['screen_fragments', 'run_docking_surrogate'],
      };
    case 'structure':
      return {
        allowed: ['map_to_structure', 'inspect_secondary_structure', 'extract_pockets'],
        blocked: ['run_docking_surrogate'],
      };
    case 'pocket':
      return {
        allowed: ['extract_pockets', 'map_pharmacophore_features', 'search_pocket_vectors'],
        blocked: ['generate_report'],
      };
    case 'screening':
      return {
        allowed: ['screen_fragments', 'run_docking_surrogate', 'map_pharmacophore_features'],
        blocked: [],
      };
    case 'report':
      return {
        allowed: ['generate_report', 'export_results'],
        blocked: ['extract_pockets', 'screen_fragments', 'run_docking_surrogate'],
      };
  }
};

/**
 * Check whether a tool may be invoked given the current phase policy.
 * blockedTools takes priority; if allowedTools is non-empty only listed tools pass.
 */
export function canInvokeTool(toolName: string, allowedTools: string[], blockedTools: string[]): boolean {
  if (blockedTools.includes(toolName)) return false;
  if (allowedTools.length > 0) return allowedTools.includes(toolName);
  return true;
}

File: agent/orchestration/tool_policy.py
# Backend Orchestration Layer — Tool Policy
"""Phase-aware tool gating policy. Maps each discovery phase to the set of
tools that are allowed and blocked at that stage.

This mirrors the frontend `toolPolicyForPhase` logic so that backend
enforcement is consistent with what the UI displays.
"""

from __future__ import annotations

from agent.orchestration.models import DiscoveryPhase


# ---------------------------------------------------------------------------
# TOOL_POLICY: canonical mapping from phase → allowed/blocked tool lists
# ---------------------------------------------------------------------------

TOOL_POLICY: dict[DiscoveryPhase, dict[str, list[str]]] = {
    DiscoveryPhase.RESIDUE: {
        "allowed": [
            "get_residue_state",
            "get_high_uncertainty_residues",
            "search_residues",
            "highlight_residues",
            "set_metric",
        ],
        "blocked": [
            "extract_pockets",
            "screen_fragments",
            "run_docking_surrogate",
            "generate_report",
        ],
    },
    DiscoveryPhase.TOPOLOGY: {
        "allowed": [
            "get_graph_metrics",
            "find_graph_bridges",
            "get_shortest_paths",
            "highlight_residues",
            "search_residues",
        ],
        "blocked": [
            "screen_fragments",
            "run_docking_surrogate",
        ],
    },
    DiscoveryPhase.STRUCTURE: {
        "allowed": [
            "get_residue_state",
            "get_allosteric_sites",
            "compare_wt_mutant",
            "highlight_residues",
        ],
        "blocked": [
            "run_docking_surrogate",
        ],
    },
    DiscoveryPhase.POCKET: {
        "allowed": [
            "get_allosteric_sites",
            "search_residues",
            "highlight_residues",
            "annotate_structure",
        ],
        "blocked": [
            "generate_report",
        ],
    },
    DiscoveryPhase.SCREENING: {
        "allowed": [
            "screen_fragments",
            "run_docking_surrogate",
            "highlight_residues",
            "annotate_structure",
        ],
        "blocked": [],
    },
    DiscoveryPhase.REPORT: {
        "allowed": [
            "generate_report",
            "export_structure_data",
            "get_provenance_lineage",
        ],
        "blocked": [
            "screen_fragments",
            "run_docking_surrogate",
        ],
    },
}


def tool_policy_for_phase(phase: DiscoveryPhase) -> dict[str, list[str]]:
    """Return the tool policy (allowed + blocked lists) for a given phase.

    Returns a dict with keys "allowed" and "blocked", each containing a list
    of tool name strings.

    Raises KeyError if the phase is not recognized (should not happen with
    the enum, but defensive).
    """
    return TOOL_POLICY[phase]
